src/ase/cw/model/Server.java

```diff
@@ -151,51 +155,92 @@ private void logAction(String string) {
+            while (!stopThread) {
+                Order currentOrder=null;
+
+                //Server is free until the server takes a order
                 setStatus(ServerStatus.FREE);
+
                 //We actually do not need this synchronized block, since our implemented orderQueue is already thread safe.
                 //But to make things more robust and consistent(It is possible to pass a non thread safe queue to the Server, in this case we would need the synchronized block)
-                try {
-                    //Wait forever until an external interruption occurs
-                    currentOrder = orderQueue.take();
-                    setStatus(ServerStatus.BUSY);
-                } catch (InterruptedException e) {
+
+                boolean takeNextOrder=true;
+                while(takeNextOrder){
+                    try {
+                        //Wait forever until an external interruption occurs
+                        currentOrder = orderQueue.take();
+                        pauseIfneeded();
+                        takeNextOrder=false;
+                    } catch (InterruptedException e) {
+                        //If interrupt happens, we know we should either pause or stop
+                        if(stopThread) {
+                            //Stop
+                            break;
+                        } else {
+                            pauseIfneeded();
+                            //If pause is done, take next order
+                            takeNextOrder = true;
+                        }
+                    }
+                }
+                if(stopThread){
                     break;
                 }
+
+                setStatus(ServerStatus.BUSY);
+
+                if(currentOrder==null){
+                    //Should never happen
+                    throw new java.lang.IllegalStateException("Current order is null");
+                }
                 //Tell listener, we proceed a new order
                 orderHandler.orderTaken(currentOrder, Server.this);
 
                 List<OrderItem> items = currentOrder.getOrderItems();
                 for (OrderItem orderItem : items) {
                     //Proceed item
                     orderHandler.itemTaken(currentOrder, orderItem, Server.this);
+
                     try {
                         Thread.sleep(processTime);
                     } catch (InterruptedException e) {
-                        //Finish the current order, then stop
-                        stopThread = true;
+                        //If interrupt happens, we know we should either pause or stop
+                        pauseIfneeded();
+                        //If the server should stop, continue finishing the order and then stop
                     }
                     //Item finished after processTime ms
                     orderHandler.itemFinished(currentOrder, orderItem, Server.this);
+                    pauseIfneeded();
```

Comment:
this is not right - server should only be paused when order is processed, not the single item